diploma-project/pages/_app.jsx

```diff
@@ -1,24 +1,64 @@
-import '../styles/globals.css'
-import {useEffect, useState} from "react";
-import {getAuth} from "firebase/auth";
-import auth from "../firebase";
+import '../styles/globals.css';
+import { useEffect } from 'react';
+import { getAuth } from 'firebase/auth';
+import { onSnapshot, query, collection, where } from 'firebase/firestore';
+import auth, { firestore } from '../firebase';
+import { useUserStore } from '../store/userStorage';
+import {getUser, User} from '../services/user';
 
 const MyApp = ({ Component, pageProps }) => {
-  const [user, setUser] = useState(null);
+  const { setUser: setUser } = useUserStore((state) => ({setUser: state.setUser}));
 
   useEffect(() => {
-    const unsubscribe = getAuth(auth).onAuthStateChanged((user) => {
-      if (user) {
-        setUser(user);
+    let unsubscribeSnapshot = null
+    const unsubscribeAuth = getAuth(auth).onAuthStateChanged((currentUser) => {
+      if (currentUser) {
+        // Store the user data in the global state
+        getUser(currentUser.email).then((userData) => {
+          setUser(userData);
+          currentUser.getIdToken().then((token) => {
+            sessionStorage.setItem('accessToken', token);
+          });
+        });
+
+        // Fetch user data from Firestore using onSnapshot
+        const q = query(collection(firestore, 'users'), where('uid', '==', currentUser.uid));
+        unsubscribeSnapshot = onSnapshot(q, (querySnapshot) => {
+          let userData = null;
+
+          querySnapshot.forEach((doc) => {
+            const data = doc.data();
+            userData = {
+              name: data.name,
+              email: data.email,
+              photoURL: data.photoURL,
+              uid: data.uid,
+              interests: data.interests,
+              requests: data.requests,
+              craft: data.craft,
+              orders: data.orders,
+              catalog: data.catalog,
+            };
+          });
+
+          setUser(userData);
+        });
       } else {
-        setUser(null);
+        setUser(new User());
+        sessionStorage.setItem('accessToken', '');
+        if (unsubscribeSnapshot) {
+          unsubscribeSnapshot();
+        }
       }
     });
 
-    return () => unsubscribe();
+    // Return a cleanup function to unsubscribe from the AuthStateChanged listener
+    return () => {
+      unsubscribeAuth();
```

Comment:
Not a 100% sure on this one, but it seems logical in the cleanup to unsubscribe from the onSnapshot as well (if it exists)